### DISCUSSION TITLE
should eBPF for Windows allow multi-level nesting in map-in-map?

### DISCUSSION THREAD

**Discussion by shpalani** [microsoft/ebpf-for-windows#2966] [Q&A]
### Describe the bug

**Request:** ebpf-for-windows should not allow multi-level nesting for map_of_maps. The inner map should not be another map-in-map.

**Example of multi-level nesting:**
**Outer map:** BPF_MAP_TYPE_ARRAY_OF_MAPS/BPF_MAP_TYPE_HASH_OF_MAPS
**Inner map**:  BPF_MAP_TYPE_ARRAY_OF_MAPS/BPF_MAP_TYPE_HASH_OF_MAPS

**Compatibility with Linux eBPF:**
As per Linux doc https://docs.kernel.org/bpf/map_of_maps.html
  - Multi-level nesting is not supported.

### OS information

Windows 10 and above

### Steps taken to reproduce bug

Map-in-map sample:
```
struct {
		__uint(type, BPF_MAP_TYPE_HASH_OF_MAPS);
		__uint(key_size, sizeof(__u32));
		__uint(value_size, sizeof(__u32));
		__uint(max_entries, 1);
} LB6_MAP_INNER  SEC(".maps");

struct {
	__uint(type, BPF_MAP_TYPE_HASH_OF_MAPS);
	__type(key, __u16);
	__type(value, __u32);
	__uint(pinning, LIBBPF_PIN_BY_NAME);
	__uint(max_entries, 65535);
	__uint(map_flags, CONDITIONAL_PREALLOC);
	/* inner map definition */
	__array(values, LB6_MAP_INNER);
} LB6_MAP_OUTER SEC(".maps");
```

### Expected behavior

bpf2c should flag an error, to unsupport more than 1 level of nesting.

### Actual outcome

Currently, multi-level nesting is allowed.

**Comment by shankarseal** [comment]
It must support 1 level of nesting though.

**Comment by shankarseal** [comment]
At least investigation should be done on if verifier can handle 1+ level of nesting.

**Comment by shpalani** [comment]
1. JFYI, Linux eBPF code which disallows:
https://github.com/torvalds/linux/blob/58720809f52779dc0f08e53e54b014209d13eebb/kernel/bpf/map_in_map.c#L22
